Enable support for Social Link block in the Email Editor

triple0t opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes #58022 WOOPLUG-4294

This PR implements support for the Social Links block in the WooCommerce Email Editor. The implementation includes:

- Support for rendering Social Link and Social Links blocks in the Email Editor
- Integration with core WordPress social link services
- Support for various social media platforms including Facebook, Twitter, Instagram, LinkedIn, and many more
- Customizable styling options including:
  - Icon sizes (small, normal, large, huge)
  - Brand colors for each social platform
  - Pill shape style support
  - Custom background colors
- Email client compatibility improvements
- Responsive design support


(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Before | After |
| ------ | ----- |
|     New   |   <img width="667" alt="Screenshot 2025-05-21 at 3 52 33 PM" src="https://github.com/user-attachments/assets/ff9271e2-3a0f-4ca9-9f17-bb4971d607f1" /> |
| New | <img width="740" alt="Screenshot 2025-05-21 at 7 30 16 AM" src="https://github.com/user-attachments/assets/1e7ad57b-b33a-4e83-a73f-af358faa520e" /> |
| New | <img width="719" alt="Screenshot 2025-05-21 at 9 54 35 PM" src="https://github.com/user-attachments/assets/00eccfc2-9600-4f80-ad07-c16faa1d10f9" />  |
| New | <img width="678" alt="Screenshot 2025-05-21 at 11 42 59 PM" src="https://github.com/user-attachments/assets/ff9648f6-9692-4ba5-ab92-4143295742d7" /> |


<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure the Block email editor is enabled. (`/wp-admin/admin.php?page=wc-settings&tab=advanced&section=features`)
2. Edit any email template from the email listing page. (`/wp-admin/admin.php?page=wc-settings&tab=email`)
3. Add the Social Links block
4. Add various social media links
5. Test different styling options:
   - Try different icon sizes
   - Test with and without labels
   - Test with pill shape style
   - Test with custom colors
6. Preview the email to verify rendering
7. Send a test email to verify compatibility with email clients

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

triple0t committed:
Enhance social links block functionality in email editor

* Introduce a new social links block with preset variations for popular platforms.
* Unregister unsupported social link variations and disable the icon color picker for the social links block.
* Update the block initialization to include the new enhancements.

This update improves the usability and customization of social links within the email editor.

triple0t committed:
Enable social links block in the email editor

triple0t committed:
Add social link and social links block renderers to the email editor

* Introduce new classes for rendering individual social link and multiple social links blocks.
* Update the block initialization to include these new renderers, enhancing the email editor's functionality for social sharing options.

triple0t committed:
Add social media icons for email editor

* Introduce new icon files for Facebook, GitHub, Instagram, LinkedIn, Twitter, WordPress, and X in various styles (black, brand, white).
* Enhance the visual options available for social links in the email editor, improving customization for users.

triple0t committed:
Refactor social links rendering and add utility functions

* Introduce Social_Links_Helper class to provide utility functions for detecting whiteish colors.
* Update Social_Links class to use instance methods instead of static methods for better object-oriented practices.
* Set default icon color to white and adjust background color based on icon color detection, enhancing visual consistency in the email editor.

triple0t committed:
Enhance social links rendering with pill shape style support

* Add support for 'is-style-pill-shape' class to adjust anchor styles, including padding and border-radius.
* Update image dimensions for better visual consistency in the social links block.
* Refactor CSS styles for table data elements to ensure proper alignment and padding.

triple0t committed:
Enhance social links block functionality in email editor

* Add BlockControls to manage additional settings for social links.
* Update type definitions for better type safety in the social links block.
* Refactor CSS styles for improved customization and visual consistency.
* Ensure proper handling of icon color and background color based on user-defined settings.

triple0t committed:
Fix the styles for social links element wrapper. Background color, padding, border, etc now works

triple0t committed:
Implement service brand color support in social links rendering

* Add a new method in Social_Links_Helper to retrieve brand colors for various services.
* Update Social_Links class to utilize the service brand color for icon background and text color, enhancing visual consistency based on service branding.

triple0t committed:
Fix broken social icon images and ensure they render well on email clients

triple0t committed:
Enhance social links functionality by adding support for additional platforms and improving URL handling

* Reintroduce previously commented-out social media platforms in the email editor's social links block.
* Implement URL validation to prepend `mailto:` for email addresses and `https://` for URLs lacking a scheme.
* Add new icon files for Behance, Bluesky, Chain, Discord, and others to improve visual representation in the email editor.

triple0t committed:
Add support for size handling in the social links block

* Remove unnecessary BlockControls for size management in the email editor's social links block.
* Introduce new methods in Social_Links_Helper to define and retrieve social link sizes, enhancing customization options.
* Update rendering logic to dynamically adjust icon sizes based on user-defined settings, ensuring better visual consistency.

triple0t committed:
Add changelog files

costasovo commented on packages/php/email-editor/src/Integrations/Core/Renderer/Blocks/class-social-link.php:
```diff
@@ -0,0 +1,30 @@
+	 * @return string
+	 */
+	protected function render_content( $block_content, array $parsed_block, Settings_Controller $settings_controller ): string {
+		// Not using this for now.
```
It would be nice to add why we are not using this flow. I assume that's because the icons are rendered in the parent's block renderer, correct?

triple0t replied:
Yup. Updated

costasovo commented on packages/js/email-editor/src/blocks/core/social-links.tsx:
```diff
@@ -0,0 +1,153 @@
+			attributes: {
+				service: 'twitter',
+				url: 'https://twitter.com/WordPress',
+			},
```
I think we shouldn't add Twitter to the preset variation. Twitter was renamed to X so we could add X or just go with only two platforms.

triple0t replied:
Noted. Implemented in https://github.com/woocommerce/woocommerce/pull/58194/commits/2f9f53b7030f5b8934bea464c1457b2130aab2d2

costasovo commented on packages/php/email-editor/src/Integrations/Utils/class-social-links-helper.php:
```diff
@@ -104,4 +104,45 @@ public static function get_service_brand_color( $service_name ) {
+			'has-small-icon-size'  => __( 'Small' ),
+			'has-normal-icon-size' => __( 'Normal' ),
+			'has-large-icon-size'  => __( 'Large' ),
+			'has-huge-icon-size'   => __( 'Huge' ),
```
Can we use the translations without `woocommerce` as the second parameter here?

triple0t replied:
The method is unused and currently unnecessary. Removed in https://github.com/woocommerce/woocommerce/pull/58194/commits/2f9f53b7030f5b8934bea464c1457b2130aab2d2

costasovo reviewed (commented):
@triple0t Great job! I think we are almost there 💪 

I left a few comments and found a couple of minor rendering issues when testing.

| Editor  | Rendered output |
| ------------- | ------------- |
| <img width="749" alt="Screenshot 2025-05-22 at 15 42 41" src="https://github.com/user-attachments/assets/bb17ffb1-bce0-4397-9a4e-485d6662a7a0" /> | <img width="739" alt="Screenshot 2025-05-22 at 15 42 13" src="https://github.com/user-attachments/assets/d634f271-0815-4123-8bd6-5be0477ddbe0" /> |


### Issues:
1) In the editor the gaps between icons both vertical and horizontal respect Styles > Layout > Block spacing setting. VS in renderd output the horizontal gap is fixed and vertical is none. I think we should unify it and we can also decide whether we want to change the appearance in renderer or in the editor.

2) The purple icons with text on the right are Default style with text and the internal padding of the icons in the editor is different compared to the rendered. The editor version is IMO nicer.

<img width="297" alt="Screenshot 2025-05-22 at 13 28 07" src="https://github.com/user-attachments/assets/bb6294b9-bc21-430d-a254-3fb1d8587214" />

3) There is a weird artefact ⬆️  in the Colors section when Logos only is active. 

4) We can optimize the PNG images used for the icons. I tested a couple of random icons and I was able to reduce their size by 40% (using https://imageoptim.com/).

5) The telegram icon image is 10 times bigger than other icons. 

6) I would recommend adding an integration test for the renderer. It is packed with logic and handles many use cases. I can imagine we can easily break something with a little change.

### Testing in Email Clients
Here is the link to EmailOnAcid project I used https://app.emailonacid.com/app/editor/project/MTm4Y7h77SOwcA4QRlfWVAPY6duzMZxYc6KKvXC43wmOM

It looks good in the most email clients.  But Outlook :( 


| Google (correctly rendered)  | Outlook |
| ------------- | ------------- |
| <img width="949" alt="Screenshot 2025-05-22 at 19 32 20" src="https://github.com/user-attachments/assets/b9667fc1-637d-4b40-99db-4fbee0214e97" /> | <img width="837" alt="Screenshot 2025-05-22 at 19 34 42" src="https://github.com/user-attachments/assets/99d36a46-73e8-4c60-8d33-177af92cbf9c" /> |

The rounded missing rounded corners are expected. I think we should address:
1) Icons in Outlook are not rendered on one line.
2) The text in labels should be bigger in the icons with texts.

triple0t committed:
Fix social links style issues

We switched to em unit due to the well supported nature in email clients https://www.caniemail.com/features/css-unit-em/

* Update CSS for social links block to improve icon color settings and visual consistency.
* Adjust font size calculation for icons and modify padding for better layout.
* Enhance HTML structure for service labels to ensure proper alignment and spacing.

triple0t committed:
Update social links to reflect rebranding and remove unused size options

* Change Twitter service name to 'x' and update the URL accordingly.
* Remove the deprecated method for retrieving social link size options from the Social_Links_Helper class, streamlining the codebase.

triple0t committed:
Optimize and reduce social icon image sizes

triple0t committed:
Fix social icons display on Outlook email clients

triple0t committed:
Add default alignment and style for social links block

* Set default alignment to 'center' and apply 'is-style-logos-only' class for improved visual consistency across email clients.

triple0t committed:
Fix PHPStan errors

triple0t committed:
Add integration and unit tests for Social_Links and Social_Links_Helper classes

triple0t commented:
Thank you for the in-depth review, @costasovo 

New commits start from https://github.com/woocommerce/woocommerce/pull/58194/commits/a1e7698484a651b75a7021f34694a4ef9b9d58dd

___

> In the editor the gaps between icons both vertical and horizontal respect Styles > Layout > Block spacing setting. VS in renderd output the horizontal gap is fixed and vertical is none. I think we should unify it and we can also decide whether we want to change the appearance in renderer or in the editor.

![Screenshot 2025-05-27 at 5 28 41 PM](https://github.com/user-attachments/assets/b35eef2c-642d-4180-851e-1ff3733b2ced)


I'm not quite sure about this. Unlike in the [post editor,](https://github.com/user-attachments/assets/73093fe9-c4c9-4b7e-8a70-27a9ec8b1f93) we don't support justification in the email editor. The align option works well in the [rendered](https://github.com/user-attachments/assets/209fa106-d4b4-48e5-b1d6-512abf289f00) version.


> The purple icons with text on the right are Default style with text and the internal padding of the icons in the editor is different compared to the rendered. The editor version is IMO nicer.

Fixed in https://github.com/woocommerce/woocommerce/pull/58194/commits/a1e7698484a651b75a7021f34694a4ef9b9d58dd

> There is a weird artefact ⬆️ in the Colors section when Logos only is active.

Fixed in https://github.com/woocommerce/woocommerce/pull/58194/commits/a1e7698484a651b75a7021f34694a4ef9b9d58dd

> We can optimize the PNG images used for the icons. I tested a couple of random icons and I was able to reduce their size by 40% (using https://imageoptim.com/).

Implemented in https://github.com/woocommerce/woocommerce/pull/58194/commits/1d16ec4f720fc12c653425f7bf11d745aaeee24c

> The telegram icon image is 10 times bigger than other icons.

Fixed in https://github.com/woocommerce/woocommerce/pull/58194/commits/1d16ec4f720fc12c653425f7bf11d745aaeee24c

> I would recommend adding an integration test for the renderer. It is packed with logic and handles many use cases. I can imagine we can easily break something with a little change.

Implemented in https://github.com/woocommerce/woocommerce/pull/58194/commits/b1b8ad25cb6578a95aa24ae1d896d5936be791da

> Icons in Outlook are not rendered on one line.

Fixed in https://github.com/woocommerce/woocommerce/pull/58194/commits/d0ac4e528d4fb1c78bdfd6edf286ce139b373abc

![Screenshot 2025-05-27 at 5 33 18 PM](https://github.com/user-attachments/assets/c3cb93a7-a390-4d98-aeb9-dbd2fcfdd54d)

Unfortunately, Outlook does not support the [border-radius](https://www.caniemail.com/features/css-border-radius/) property. The [alternative](https://learn.microsoft.com/en-us/windows/win32/vml/msdn-online-vml-roundrect-element) is deprecated. Investing time in developing a deprecated format doesn't make much sense. We can live with the box corner.

> The text in labels should be bigger in the icons with texts.

Implemented in https://github.com/woocommerce/woocommerce/pull/58194/commits/a1e7698484a651b75a7021f34694a4ef9b9d58dd.

costasovo commented on packages/php/email-editor/src/Integrations/Utils/class-social-links-helper.php:
```diff
@@ -130,4 +130,23 @@ public static function get_social_link_size_option_value( $size ) {
+	 * @param string $needle The substring to search for.
+	 * @return bool True if $haystack starts with $needle, false otherwise.
+	 */
+	public static function str_starts_with( string $haystack, string $needle ): bool {
```
I noticed that function in the previous review, but when checking it, I found that WordPress has a polyfill for that. See https://github.com/WordPress/WordPress/blob/5e1454dd91283a8f356c3258eba13ccc23753a85/wp-includes/compat.php#L394-L414
So this looks like a PHPStan config issue, and we shouldn't need to add the polyfill ourselves.
In case it is an issue in Unit tests, we may need to move it to integration tests, where WP is loaded.

triple0t replied:
Yeah, this was due to PHPStan. Fixed in https://github.com/woocommerce/woocommerce/pull/58194/commits/7f401761ff5bc97fe7d881df12585380a4352ba8

costasovo reviewed (commented):
@triple0t Thank you for addressing the issues! Great job! Especially addressing the Outlook and adding the test coverage.

I added a comment about the polyfill that is not necessary, and I did another round of testing and found a few issues. Nothing major, though. Most of the things work as expected!

> Unfortunately, Outlook does not support the [border-radius](https://www.caniemail.com/features/css-border-radius/) property. The [alternative](https://learn.microsoft.com/en-us/windows/win32/vml/msdn-online-vml-roundrect-element) is deprecated. Investing time in developing a deprecated format doesn't make much sense. We can live with the box corner.

This is perfectly fine. We don't support border radius for Outlook.

> >In the editor the gaps between icons both vertical and horizontal respect Styles > Layout > Block spacing setting. VS in renderd output the horizontal gap is fixed and vertical is none. I think we should unify it and we can also decide whether we want to change the appearance in renderer or in the editor.

> I'm not quite sure about this. Unlike in the [post editor,](https://github.com/user-attachments/assets/73093fe9-c4c9-4b7e-8a70-27a9ec8b1f93) we don't support justification in the email editor. The align option works well in the [rendered](https://github.com/user-attachments/assets/209fa106-d4b4-48e5-b1d6-512abf289f00) version.

I think I was not clear in the description. The issue is still present. The issue is that the editor renders the gap between icons based on the styles settings. I recorded a video to illustrate the issue.


https://github.com/user-attachments/assets/59a69a24-5fd2-453b-a463-8a5b31936c50

I think we should address this. We could use the value from the style settings and apply half of it as a padding on the individual icon wrappers. Alternatively, we could alter the CSS for the editor to prevent the gap between icons, thereby achieving a WYSIWYG experience.

### Background-related issues
<img width="1052" alt="Screenshot 2025-05-28 at 10 48 31" src="https://github.com/user-attachments/assets/4ca98694-c5fe-464a-a7a9-db471317b086" />

In the image above, focus on the icons with the green background. On the left is the editor, and on the right is the rendered output. There are two issues:

1) The icons in the editor are not centered correctly. It happens when a background is applied to the social icons block. I found the CSS causing this is: `ul.has-background, ol.has-background { padding-left: 40px; }`

2) There is a slight gap/padding below icons in the rendered output. See the green line below the icons in the rendered output. It is not padding, but this issue is present in TD when some whitespaces are present. It can be eliminated by setting font-size: 0px; on the wrapper. I see that we set the font size for the label explicitly, so I think we can use the font-size: 0px trick as a fix here.

### Label alignment

<img width="772" alt="Screenshot 2025-05-28 at 9 26 53" src="https://github.com/user-attachments/assets/00ceed1e-5ac8-481c-9fdd-34724b7026d8" />

You can notice that the labels in the purple icons are slightly off vertically. The TD that wraps the icon has alignment set correctly to `middle`. The problem here is that font size is applied to the nested anchor element and in that case the vertical alignment doesn't work properly. It can be fixed by moving the font size from the anchor to the TD as illustrated by the arrow on the image.

triple0t committed:
Fix styling issues and ensure support for WYSIWYG experience.

triple0t committed:
Refactor Social_Links class to use native str_starts_with function and remove polyfill from Social_Links_Helper. Update PHPStan configuration to ignore errors related to str_starts_with for PHP 7.4 compatibility.

triple0t committed:
Enhance WYSIWYG experience by overriding background color for social links block in the editor. This change ensures better visibility for icons without a default background color.

triple0t committed:
Remove !important from padding-left in social links block CSS for improved styling consistency in WYSIWYG editor.

costasovo reviewed (approved):
@triple0t Thank you for the fixes! Great job! I think it is ready for merge!

> I think we can live with this as the issue is not highly perceivable. Also, the background color is not really usable without some padding, so there's no need to change anything here.

Agree. I thought it might be an easy fix, but it is not worth complex changes.

### Testing performed

I downloaded the zip from the branch and tested on a jurrasic ninja site.

| Editor | Renderer |
| ------ | --------- |
| ![Screenshot 2025-05-29 at 15 13 31](https://github.com/user-attachments/assets/4fba03e5-1f98-46b0-b531-a6e7b6d2aa24) |  ![Screenshot 2025-05-29 at 15 13 38](https://github.com/user-attachments/assets/6c556da1-7fd2-482f-80df-af080bcea05f) |

I also tested with [the Email On Acid service](https://app.emailonacid.com/app/editor/project/MTm4Y7h77SOwcA4QRlfWVAPY6duzMZxYc6KKvXC43wmOM):

![Screenshot 2025-05-29 at 15 17 13](https://github.com/user-attachments/assets/5d44f883-fce3-4203-888d-8fb76fa600d4)
![Screenshot 2025-05-29 at 15 17 33](https://github.com/user-attachments/assets/e6b037e1-393b-4eeb-b737-92b1f5b3ace2)
![Screenshot 2025-05-29 at 15 18 03](https://github.com/user-attachments/assets/680e99c0-ea5f-437f-952d-9a33a1c162b7)
![Screenshot 2025-05-29 at 15 18 53](https://github.com/user-attachments/assets/2b109e02-adba-4026-a01f-f5a58969ceb7)
